roofit/roofitcore/src/RooFuncWrapper.cxx

```diff
@@ -136,3 +165,53 @@ void RooFuncWrapper::gradient(const double *x, double *g) const
 
    _grad(const_cast<double *>(x), _observables.data(), g);
 }
+
+void RooFuncWrapper::buildCode(RooAbsReal const &head, std::string &code, RooArgSet const & /* paramSet */,
+                               RooArgSet const &ObsSet, const RooAbsData *data)
+{
+   RooFit::Detail::CodeSquashContext ctx;
+
+   // First update the result variable of params in the compute graph to in[<position>].
+   int idx = 0;
+   for (auto *param : static_range_cast<RooRealVar *>(_params)) {
+      ctx.addResult(param, "params[" + std::to_string(idx) + "]");
+      idx++;
+   }
+
+   // Also update observables...
+   idx = 0;
+   if (!ObsSet.empty()) {
+      auto dataSpans = data->getBatches(0, data->numEntries());
+      _observables.reserve(_observables.size() * data->numEntries());
+      for (auto *obs : static_range_cast<RooRealVar *>(ObsSet)) {
+         RooSpan<const double> span{dataSpans.at(obs)};
+         // If the observable is a scalar, set its name to the start index.
+         // else, store the start index and later set the the name to obs[start_idx + curr_idx], here curr_idx is
+         // defined by a loop producing parent node.
+         if (data->numEntries() == 1)
+            ctx.addResult(obs, "obs[" + std::to_string(idx) + "]");
+         else
+            ctx.addVecObs(obs, idx);
+         idx += data->numEntries();
+      }
+   }
+
+   // This will not work for nodes that produce loops as we need to keep track of the subtree of the loop producing
+   // node. A better approach is to have 2 stacks and perform an iterative post order traversal on the graph/resolved
+   // tree.
+   RooArgSet nodes;
+   RooHelpers::getSortedComputationGraph(head, nodes);
+
+   for (RooAbsArg *node : nodes) {
+      RooAbsReal *curr = dynamic_cast<RooAbsReal *>(node);
+      if (!curr) {
+         std::stringstream errorMsg;
+         errorMsg << "Translate is only supported for RooAbsReal derived objects.";
+         oocoutE(nullptr, Minimization) << errorMsg.str() << std::endl;
+         throw std::runtime_error(errorMsg.str().c_str());
+      }
+      code += curr->translate(ctx);
```

Comment:
There is something sketchy going on here. You are building the code with these calls to `translate`, appending it to the variable `std::string code`. But then later on, you overwrite the string in the call to `ctx.assembleCode(ctx.getResult(&head), code)`, where `code` is an output parameter.

That means the return value of `translate` is actually not necessary. Indeed, I think it's better if `translate` returns `void`, because all the information flows through the context class.

What do you think about this?

Reply:
It is absolutely necessary for the translate function to return a string. The assemble function is only to add the global statements and the return statement. Sure you could change the function to not overwrite and return the string instead.

Reply:
I don't understand how it is "absolutely necessary" if it always returns an empty string and the tests pass just fine?

Reply:
Ah okay I think I get it now. It will be important later for the NLL

Reply:
Yes exactly, we only have super simple cases here and that's why it's not obvious why we need translate to return a string.